tools/html_transformer/html_transformer.js

```diff
@@ -0,0 +1,105 @@
+var Promise = require('es6-promise').Promise;
+var jsdom = require('jsdom');
+
+exports.load = function(filename) {
+  return new Promise(function(resolve, reject) {
+    jsdom.env(filename, function(errors, window) {
+      if (errors && errors.length) {
+        return reject(errors[0]);
+      }
+
+      return resolve(new HTMLTransformer(window));
+    });
+  });
+};
+
+function HTMLTransformer(window) {
+  this.window = window;
+}
+exports.HTMLTransformer = HTMLTransformer;
+
+HTMLTransformer.prototype = {
+  getSharedDependencies: function() {
+    var doc = this.window.document;
+    var links = getSharedElements(doc, 'link');
+    var scripts = getSharedElements(doc, 'script');
+    var commented = flatmap(getComments(doc), getSharedElementsFromComment);
+    return links.concat(scripts).concat(commented);
+  }
+};
+
+function getSharedElements(doc, selector) {
+  return querySelectorAll(doc, selector).map(getShared).filter(function(x) {
+    return !!x;
+  });
+}
+
+function getSharedElementsFromComment(comment) {
+  var sharedLink = /<link\s.*href="(app:\/\/.+\.gaiamobile\.org)?\/?(shared\/.+)".*>/;
+  var sharedScript = /<script\s.*src="(app:\/\/.+\.gaiamobile\.org)?\/?(shared\/.+)".*>/;
+
+  return flatmap(comment.split(/\n/), function(commentLine) {
+    var links = commentLine.match(sharedLink) || [];
+    var scripts = commentLine.match(sharedScript) || [];
+    var results = links.concat(scripts);
+    return results.filter(function(result) {
+      return result && result.indexOf('shared') === 0;
+    });
+  });
+};
+
+// Neither document.createTreeWalker nor document.createNodeIterator
+// are supported by jsdom so we're going to do this the old fashioned way...
+function getComments(doc) {
+  var results = [];
+  var queue = [doc.body];
+  while (queue.length) {
+    var next = queue.shift();
```

Comment:
Okay, I've done some children walking like this with for + function calls (not recursion). But I think in this code your expression is clear and small enough, so I think it's fine for me to keep it here.